### ISSUE TITLE
Time based queries

### ISSUE THREAD

**Issue by mrosenberg** [tick/tick-api#35]
Is it possible or could it be made possible to query projects and tasks by time? We are looking at building reports that span multiple timeframes. If we query entries that include closed projects we have to download an awful lot of data we don't actually need.

**Comment by tomrossi7** [comment]
Hi Matthew! How are you thinking about querying projects and tasks by time? Are you looking for projects that have a certain amount of time charged to them?

Using the updated_at attribute with the /entries endpoint lets you query only entries after the last time you pulled down data. Depending on what you are trying to do that may help you to narrow down the amount of data you end up pulling back.

**Comment by mrosenberg** [comment]
We weren't really looking to download the info on a regular basis.

Say I wanted too look at the productivity of November 2016 to compare with November 2017. I can query the entries for those dates but I need more information to present any useful info. What tasks took the most? What project drew the most resources? Which client did we expend the most effort on? 

To answer those questions for any given time period I have to query the entries then query each task by id and save the results. That could potentially be almost 5,000 http requests. 

Tasks and projects both have a created_at and date_closed attributes. If I could query those the same way you can entries I could get all the necessary data in 3 requests.

**Comment by tomrossi7** [comment]
I'm not sure I follow. If you wanted to compare productivity from November 2016 to November 2017, you would need all of the time entry data for both months. How would querying projects or tasks created_at and date_closed help?

**Comment by mrosenberg** [comment]
If I query the entires endpoint 

`/api/v2/entries.json?start_date=2017-11-01&end_date=2017-11-30`

I get back an array of entry information.

```
[
    {
        "id": xxxxxxxxx,
        "date": "2017-11-09",
        "hours": 0.3667,
        "notes": "",
        "task_id": xxxxxxx,
        "user_id": xxxxxx,
        "url": "http://secure.tickspot.com/xxxxx/api/v2/entries/xxxxx.json",
        "created_at": "2017-11-09T15:16:49.000-05:00",
        "updated_at": "2017-11-09T15:16:49.000-05:00"
    },
......
]
```

That doesn't tell me what task, project, client, or user made that entry without inquiring further. If the projects are open, I can hit up the tasks.json and projects.json endpoints and get the names to put with the IDs included in each entry. After the projects are closed, the projects/closed.json endpoint returns way more info than I need. Same thing for tasks.

**Comment by tomrossi7** [comment]
Whether the projects are opened or closed, you should be able to make the one API call to get the corresponding task (e.g. /tasks/1234). If it were me, I would download all the entries for the time ranges and then grab the unique task_ids and make all of those API requests. That would give me everything I need to compare productivity from one month to the other. Does that make sense?

I'm not sure how adding query options to projects and tasks is going to reduce the number of API calls?

**Comment by mrosenberg** [comment]
I considered going that route. For last month, that would require me to make 782 requests to the task endpoint.

**Comment by tomrossi7** [comment]
On how many entries?

**Comment by mrosenberg** [comment]
3,421. That's for one month.

**Comment by tomrossi7** [comment]
So, we decided early on in the API design that **we prefer more-fast-smaller API requests rather than fewer-slower-large requests**.

So in this particular case we chose not to make the one '/entries' request a lot larger by including task/project information for each of the 3421 entries and instead requiring 782 additional small fast requests to get the additional information. Neither approach is right or wrong since they both have trade-offs -- its just a design choice we made. I hope that helps at least explain why things work the way they do!

**Comment by mrosenberg** [comment]
That's a completely sensible decision. I think adding similar date queries to the other endpoints would give you the best of both worlds, and support power users of the your platform.

**Comment by tomrossi7** [comment]
Yeah, that is what I was trying to understand. I'm not sure how adding the ability to query projects and tasks based on created/updated is going to help? Do you have an example?

**Comment by mrosenberg** [comment]
Assuming all older info, say December of 2016. I can query the entries endpoint for all items dated between 2016-12-01 and 2016-12-31. All those entries would be associated with tasks and projects created on 2016-12-01. If could could query the created_on date for projects and tasks, I would get just the ones I need in a single request. 

For example if I query projects.json I get a list of all the projects for the current month. I'd like to get that list for previous months as well. Query the projects/closed.json gives me waaay more than I need. 

For instance:
`/api/v2/projects.json?start_date=2016-12-01&end_date=2016-12-31`

Would return all projects created December of 2016. Same for tasks.

**Comment by tomrossi7** [comment]
> All those entries would be associated with tasks and projects created on 2016-12-01.

No, the time entries from December 2016 could be for projects that were created months before?\

> For example if I query projects.json I get a list of all the projects for the current month.

No, you get a list of all the projects that are currently open?

>  I'd like to get that list for previous months as well. 

Are you saying you would like to get a list of projects that were opened during a particular month? We don't store the data necessary to generate that list.

**Comment by mrosenberg** [comment]
For our use case, that would be very helpful. We create projects for every month, for every client.
